Adds Prisma client generation during app generation

shayneczyzewski opened the pull request:
# Description

Currently, we do not regenerate the Prisma client as part of our overall app generation process. This means if the user makes changes to their Wasp files that impact schema.prisma, those changes won't be applied to the Prisma client until they run `wasp db migrate`. This is not ideal, and it would be much better if we made this a part of normal app generation, in line with the Prisma recommendation:

> Important: You need to re-run the prisma generate command after every change that's made to your Prisma schema to update the generated Prisma Client code.
Ref: https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/generating-prisma-client

## Important

The Prisma client generation step runs via `npx`, and is now run on whenever the prisma schema changes (and thus differs from schema.prisma.wasp-generate-checksum). One part of this we needed to solve was for this to work even if Prisma was not installed on the user's machine yet. This is because if it is not found, it prompts them to install it, and it uses the latest version. This change adds a function that tries `npx` commands that will fail if the exact version we use is not available locally, and then `npm install`s before retrying.

This process will also hide all output from `prisma generate` unless there is an error on retry.

Fixes #427

## Type of change

Please select the option(s) that is more relevant.

- [ ] Code cleanup
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

shayneczyzewski commented on waspc/src/Wasp/Generator.hs:
```diff
@@ -55,9 +56,6 @@ genApp spec =
     <++> genDockerFiles spec
 
-checkGeneratedCode :: AppSpec -> Path' Abs (Dir ProjectRootDir) -> IO [GeneratorWarning]
-checkGeneratedCode spec dstDir = DbGenerator.checkGeneratedCode spec dstDir
```
Removed the level of indirection (`DbGenerator.checkGeneratedCode`) and called the underlying function (`DbGenerator.warnIfDbSchemaChangedSinceLastMigration`) directly in `postGenDbGeneratorActions `.

shayneczyzewski commented on waspc/src/Wasp/Generator.hs:
```diff
@@ -29,7 +30,7 @@ import Wasp.Util ((<++>))
 -- | Generates web app code from given Wasp and writes it to given destination directory.
 --   If dstDir does not exist yet, it will be created.
---   If there are any errors returned, that means that generator failed and no new code was written.
+--   If there are any errors returned, that means that generator failed but new code was possibly still written.
```
Updated the semantics here as `DbGenerator.genPrismaClient` can fail _after_ code is written.

shayneczyzewski commented on waspc/src/Wasp/Generator/DbGenerator.hs:
```diff
@@ -3,75 +3,34 @@
-
-dbSchemaChecksumFileInProjectRootDir :: Path' (Rel ProjectRootDir) (File DbSchemaChecksumFile)
-dbSchemaChecksumFileInProjectRootDir = dbRootDirInProjectRootDir </> dbSchemaChecksumFileInDbRootDir
-
```
Moved all of the above into a new file: `src/Wasp/Generator/DbGenerator/Common.hs`. This was done, and a few other refactoring, to prevent cyclical dependencies and mirror other Generators.

shayneczyzewski commented on waspc/src/Wasp/Generator/DbGenerator.hs:
```diff
@@ -156,12 +115,8 @@ genMigrationsDir spec =
   where
     genProjectMigrationsDir = dbRootDirInProjectRootDir </> dbMigrationsDirInDbRootDir
 
-writeDbSchemaChecksumToFile :: Path' Abs (Dir ProjectRootDir) -> IO ()
```
Moved into `src/Wasp/Generator/DbGenerator/Operations.hs` for cyclical deps and for a better fit as to what it is doing/using.

shayneczyzewski commented on waspc/src/Wasp/Generator/DbGenerator/Jobs.hs:
```diff
@@ -27,28 +35,96 @@ migrateDev projectDir = do
+-- | Retries the original job when there is an error by first running the second job, and then
+-- retrying the first.
+retryJobOnErrorWith :: J.Job -> J.Job -> J.Job
+retryJobOnErrorWith originalJob afterErrorJob externalChannel = do
```
Open to better name here :D

shayneczyzewski commented on waspc/src/Wasp/Generator.hs:
```diff
@@ -77,3 +75,9 @@ writeDotWaspInfo dstDir = do
   let dstPath = dstDir </> [relfile|.waspinfo|]
   Data.Text.IO.writeFile (SP.toFilePath dstPath) (Data.Text.pack content)
+
+postGenDbGeneratorActions :: AppSpec -> Path' Abs (Dir ProjectRootDir) -> IO ([GeneratorWarning], [GeneratorError])
```
Not sold on the name here, but wanted a place to put DbGenerator things after gen, where one is a check that can produce warnings, and the other is an op that can produce errors.

shayneczyzewski replied:
Sounds good, I will update the name and add a comment 👍🏻

shayneczyzewski commented on waspc/src/Wasp/Generator/DbGenerator/Operations.hs:
```diff
@@ -60,3 +65,24 @@ copyMigrationsBackToSource genProjectRootDirAbs dbMigrationsDirInWaspProjectDirA
     waspMigrationsDir = SP.castDir dbMigrationsDirInWaspProjectDirAbs
     genProjectMigrationsDir = SP.castDir $ genProjectRootDirAbs SP.</> dbRootDirInProjectRootDir SP.</> dbMigrationsDirInDbRootDir
+
+writeDbSchemaChecksumToFile :: Path' Abs (Dir ProjectRootDir) -> IO ()
```
Moved here from DbGenerator.hs

shayneczyzewski commented on waspc/src/Wasp/Generator/DbGenerator/Jobs.hs:
```diff
@@ -27,28 +35,96 @@ migrateDev projectDir = do
 
   -- NOTE(martin): For this to work on Mac, filepath in the list below must be as it is now - not wrapped in any quotes.
-  let npxPrismaCmd = ["npx", "prisma", "migrate", "dev", "--schema", SP.toFilePath schemaFile]
+  let npxPrismaCmd = ["npx", "--no-install", prismaNpxCommand, "migrate", "dev", "--schema", SP.toFilePath schemaFile]
```
https://www.npmjs.com/package/npx

shayneczyzewski commented on waspc/src/Wasp/Generator/DbGenerator.hs:
```diff
@@ -156,12 +115,8 @@ genMigrationsDir spec =
-    dbSchemaChecksumFp = SP.fromAbsFile $ genProjectRootDir </> dbSchemaChecksumFileInProjectRootDir
+-- | NOTE(shayne): We cannot check for the difference between the prisma schema and checksum file here, as we may need
+-- to regnerate the client after it was "dirtied" by a change that made the checksums differ and then was undone.
+-- TODO(shayne): Any other heuristics we can use to do this less often?
```
The reason we can't diff checksum is for example:
- Change entity column, checksum diffs, we regenerate client
- Undo that change, checksum same, we do not regenerate client but it is now still the old/last/wrong change version

shayneczyzewski replied:
Therefore, I thought it was safest to just always run for now (until some better heuristic is added).

shayneczyzewski replied:
We could also add another prisma checksum file, one for last migrate, and one for last generate, and then we can check that one.

shayneczyzewski replied:
Yeah, I probably should have just pulled the trigger on that optimization myself to begin with. I will add a checksum for `generate` so we don't have to call it all the time and keep compilation speedy in the general case 👍🏻

Martinsos replied:
Yes that is it, two checksum files!

shayneczyzewski replied:
Added second checksum to speed it up! 👍🏻

shayneczyzewski commented:
@Martinsos I believe I am ready for the review now. I have done some local testing, but will continue to do more. There is one open question in the PR description it may be good to think about. Thanks!

Martinsos commented:
> 

Output form `npx prisma generate` during install -> I think we should hide it, as it can be pretty distracting at the moment. Messages about updating are confusing because they can't really update prisma, they need to update whole Wasp. The rest of instructions is also not useful. So I would do as you suggested: hide it, and if error happens, then we can print the error message.

shayneczyzewski commented:
> > 
> 
> Output form `npx prisma generate` during install -> I think we should hide it, as it can be pretty distracting at the moment. Messages about updating are confusing because they can't really update prisma, they need to update whole Wasp. The rest of instructions is also not useful. So I would do as you suggested: hide it, and if error happens, then we can print the error message.

Sounds good, thanks @Martinsos. I will update it to only show output during Prisma client generation if an error is returned. I will have to tinker a bit with the timing/mechanism of channel message forwarding, but shouldn't be too bad. 👍🏻

shayneczyzewski commented:
Updated this PR to only show output when the `ExitCode` is a failure @Martinsos 👍🏻 thanks!

shayneczyzewski commented:
One thing I am still unsure of, is should we consider adding _yet another Prisma related checksum_ (in addition to migrate) to help us determine if we should be running prisma generate on each compile? It doesn't take _too long_ to run, but definitely longer than before (maybe a couple of seconds). Or is that an optimization for later?

Martinsos reviewed (commented):
Nice job @shayneczyzewski , this is going to be a very cool feature!

It is great to see that you got "immersed" into the Jobs and Channels, I see you are finding your way through it nicely. The fact that Job is a `type` that actually takes a Channel as an input can be confusing sometimes to the reader of the code I think.

I let some comments!

Martinsos commented:
> One thing I am still unsure of, is should we consider adding _yet another Prisma related checksum_ (in addition to migrate) to help us determine if we should be running prisma generate on each compile? It doesn't take _too long_ to run, but definitely longer than before (maybe a couple of seconds). Or is that an optimization for later?

Ah I didn't see you asked this but I already commented on it in the code! Yes I would certainly add that now. We shouldn't over optimize at this point, but on the other hand we want to keep compilation short as it is part of the feedback loop during development, and it is not so much work to add this. So I would say yes let's do it!

shayneczyzewski commented on waspc/src/Wasp/Generator/DbGenerator/Jobs.hs:
```diff
@@ -1,18 +1,26 @@
 import Wasp.Generator.ServerGenerator.Common (serverRootDirInProjectRootDir)
 
+prismaNpxCommand :: String
+prismaNpxCommand = "prisma@" ++ prismaVersion
```
Good call, I will pull all of that in, and add the comment to address why we have it no install and a version! 👍🏻

shayneczyzewski commented on waspc/src/Wasp/Generator/DbGenerator/Jobs.hs:
```diff
@@ -27,28 +35,111 @@ migrateDev projectDir = do
+  retryJobOnErrorWith job (npmInstall projectDir) jobOutputChannelForwarder
+
+-- | Runs `prisma generate` to generate the Prisma client.
+runGenerate :: Path' Abs (Dir ProjectRootDir) -> J.Job
```
Yep, I will update the name to be more descriptive, even without the obvious comment above (which I just had there since a few others had similar, maybe will just nix it there too after rename)

shayneczyzewski commented on waspc/src/Wasp/Generator/DbGenerator/Jobs.hs:
```diff
@@ -27,28 +35,111 @@ migrateDev projectDir = do
+-- | Runs the original job, and if there is an error, it attempts recovery by running the second job
+-- before retrying the original one more time.
+-- It also takes a function that handles forwarding channel output.
+retryJobOnErrorWith :: J.Job -> J.Job -> ChannelForwarder -> J.Job
```
:D Thanks I appreciate it. Yeah right now it does bake in the DB job type and the forwarder logic, and is only used by DB jobs for now... hmm, will chew on moving them over and see. Thnx for spitballing it with me though 👍🏻

matijaSos replied:
it is at least localized in this Db-scoped file, so I guess no harm for now since it is not exported anywhere?

shayneczyzewski commented on waspc/src/Wasp/Generator/DbGenerator/Jobs.hs:
```diff
@@ -27,28 +35,111 @@ migrateDev projectDir = do
+-- | Runs the original job, and if there is an error, it attempts recovery by running the second job
+-- before retrying the original one more time.
+-- It also takes a function that handles forwarding channel output.
+retryJobOnErrorWith :: J.Job -> J.Job -> ChannelForwarder -> J.Job
```
I could think about pulling it higher, but the problem then is each caller would then need to remember to wrap their channel forwarder if they called this wrapper. Also, to one of the other comments in here, the reason we have to be careful with returning the `JobExit` is if we allowed the exit to flow from the job on error, and wanted to retry, the actual client would stop listening. Here is how these are typically used:

```haskell
  (_, dbSetupResult) <- liftIO (concurrently (readJobMessagesAndPrintThemPrefixed chan) (setupServer genProjectDir chan))
```

So we can only send 1 `JobExit` message, and that has to be the last thing we send. If we sent the exit from failed job 1, it would not get anything from retry.

shayneczyzewski replied:
In thinking it over, I think as a caller having to filter all but the last exit message may be messy? That was the hope of encapsulating that logic into the function.

Martinsos replied:
As a caller of `retryJobOnErrorWith`, it sounds equally complicated to me to pass the channel forwarder to it vs call the channel forwarder on it. But I didn't try coding it so I am not 100% sure.

So let's imagine we don't want to have channel forwarder at all, and we go with default behaviour, which is just forwarding all the output. How would we then want `retryJobOnErrorWith` to work? I guess we would output all the JobOutput, but would make sure to not pass that first JobExit. So basically we forward everything, but silence/ignore the JobExit from first try. The rest of JobExits we forward. Would that be ok default behaviour? We could also in theory silence the output up to that first JobExit.

So what if we had that and then on top of that wanted to do smth different, to silence all output unless there is an error -> how would we do that? Coulnd't we just wrap that channel into another one and ignore all output unless JobExit comes? How is that different than the current ChannelForwarder approach?

THere is one benefit I can see in passing the Channel forwarder to the retryJobOnErrorWIth -> if you would give it extra information, e.g. if you would let it know if it is, at the moment of specific message, in the first try, in the fallback job, or in the retry -> that information might be useful to determine what to show or not to show. For example you could choose to not output the messages from the first try if it fails, but do output the messages from the retry, even if it fails.

Am I missing the point here much, what am I overlooking?

Martinsos replied:
At the end it seems to me that it might be beneficial to pass it in like this, if we give it more control. Otherwise I might consider taking it out, just to simplify stuff -> that is what I am trying to achieve with questioning this, to possibly have simpler, individual functions.

shayneczyzewski replied:
I tried to clean this up a bit by passing in a forwarding strategy, but keeping the forwarder functions themselves local (to ensure the no JobExit message constraint). I also added a notion of first and second attempt to help the only showing prisma generate output if the command fails on retry. I _think_ this may be about as clean as we can make it, but open for feedback! Thanks

shayneczyzewski commented on waspc/src/Wasp/Generator/DbGenerator/Jobs.hs:
```diff
@@ -27,28 +35,111 @@ migrateDev projectDir = do
+            J._jobType = J.Db
+          }
+
+type ChannelForwarder = Chan JobMessage -> Chan JobMessage -> IO ()
```
Yep good call, will update

shayneczyzewski replied:
`JobMessageForwarder` maybe?

shayneczyzewski commented on waspc/src/Wasp/Generator/DbGenerator/Jobs.hs:
```diff
@@ -27,28 +35,111 @@ migrateDev projectDir = do
+  message <- readChan fromChan
+  case J._data message of
+    JobOutput _ _ -> writeChan toChan message >> jobOutputChannelForwarder fromChan toChan
+    JobExit _ -> return ()
```
From a reply above:

Also, to one of the other comments in here, the reason we have to be careful with returning the `JobExit` is if we allowed the exit to flow from the job on error, and wanted to retry, the actual original caller would stop listening. Here is how these are typically used:

```haskell
  (_, dbSetupResult) <- liftIO (concurrently (readJobMessagesAndPrintThemPrefixed chan) (setupServer genProjectDir chan))
```

So we can only send 1 JobExit message, and that has to be the last thing we send. If we sent the exit from failed job 1, it would not get anything from retry onward.

shayneczyzewski replied:
I can at least add comments to help explain this better.

Martinsos replied:
Aha, ok this is a good point, thanks for explaining!

shayneczyzewski commented on waspc/src/Wasp/Generator/DbGenerator/Jobs.hs:
```diff
@@ -27,28 +35,111 @@ migrateDev projectDir = do
+    JobOutput _ _ -> writeChan toChan message >> jobOutputChannelForwarder fromChan toChan
+    JobExit _ -> return ()
+
+errorOnlyJobOutputChannelForwarder :: ChannelForwarder
```
Good points, I will try to clean this up a bit 👍🏻

shayneczyzewski replied:
Thinking `silentUntilError :: JobMessageForwarder` and `allJobOutput :: JobMessageForwarder` above it may be ok?

shayneczyzewski commented on waspc/src/Wasp/Generator/DbGenerator/Operations.hs:
```diff
@@ -1,30 +1,37 @@
+import Wasp.Generator.Job.IO (printJobMessage, readJobMessagesAndPrintThemPrefixed)
+import Wasp.Util (checksumFromFilePath, hexToString)
 
+-- | NOTE(shayne): This appears to be used instead of readJobMessagesAndPrintThemPrefixed during migration
```
Yeah, I should move closer to the call instead 👍🏻

shayneczyzewski replied:
I may actually nix it. Was thinking it may help in case something wants to refactor it and then doesn't see the messed up output but it may not require that special attention at all.

shayneczyzewski commented on waspc/src/Wasp/Generator/DbGenerator/Operations.hs:
```diff
@@ -60,3 +67,24 @@ copyMigrationsBackToSource genProjectRootDirAbs dbMigrationsDirInWaspProjectDirA
+    dbSchemaFp = SP.fromAbsFile $ genProjectRootDirAbs SP.</> dbSchemaFileInProjectRootDir
+    dbSchemaChecksumFp = SP.fromAbsFile $ genProjectRootDirAbs SP.</> dbSchemaChecksumFileInProjectRootDir
+
+generatePrismaClient :: Path' Abs (Dir ProjectRootDir) -> IO (Either String ())
```
Sounds good! 👍🏻

shayneczyzewski commented on waspc/src/Wasp/Generator/ServerGenerator.hs:
```diff
@@ -131,7 +131,7 @@ waspNpmDeps =
       ("express", "~4.16.1"),
       ("morgan", "~1.9.1"),
-      ("@prisma/client", "2.22.1"),
+      ("@prisma/client", prismaVersion),
```
Yep, I checked, but good to always double-check! :) The only other place we use `npx prisma` where we actually do not need it is in the Dockerfile, since it runs _after_ `npm install`

shayneczyzewski commented:
Thanks for the great review @Martinsos ! 🥳 I appreciate it and will clean up the code based on the comments, as well as add the new checksum for generate so we don't always run it 👍🏻 🚀

matijaSos commented on waspc/src/Wasp/Generator/DbGenerator/Jobs.hs:
```diff
@@ -27,28 +35,111 @@ migrateDev projectDir = do
+
+-- | Runs `npm install` to install dependencies.
+-- May be needed before npx commands if dependencies (e.g. Prisma) are not installed.
+npmInstall :: Path' Abs (Dir ProjectRootDir) -> J.Job
```
hmm sometimes we use "run" as a prefix and sometimes we don't. Do we have any rule of thumb for it? Not a biggie, just thinking out loud. I am sure it will all become more standardized as we keep working on the code.

shayneczyzewski replied:
I think `runStudio` is the only db action we prefix with run atm.

matijaSos reviewed (commented):
Another PR I managed to review, woohoo :D Looks pretty good from my side, I didn't go too deep into channel stuff, mostly added some suggestions around naming/style.

shayneczyzewski commented on waspc/src/Wasp/Generator/DbGenerator.hs:
```diff
@@ -3,75 +3,34 @@
 preCleanup = deleteGeneratedMigrationsDirIfRedundant
 
-checkGeneratedCode :: AppSpec -> Path' Abs (Dir ProjectRootDir) -> IO [GeneratorWarning]
-checkGeneratedCode spec projectRootDir = maybeToList <$> warnIfDbSchemaChangedSinceLastMigration spec projectRootDir
```
Yeah this got removed and is called directly elsewhere now 👍🏻

shayneczyzewski commented:
> Another PR I managed to review, woohoo :D Looks pretty good from my side, I didn't go too deep into channel stuff, mostly added some suggestions around naming/style.

Yahoo!! 🥳 Thanks for the eyes @matijaSos!

shayneczyzewski commented:
I updated this PR addressing comments, and adding a new checksum for generate so we can be smarter about it. I also tried to refactor the tricky `retryJobOnErrorWith`. Some of it I think is just unavoidable complexity with the channel forwarding, but if anyone has further ideas on how to improve please let me know. Thank you both! CC @Martinsos @matijaSos

Martinsos commented on waspc/src/Wasp/Generator/DbGenerator/Common.hs:
```diff
@@ -0,0 +1,62 @@
+
+-- | This file represents the checksum of schema.prisma
+-- at the point at which `prisma db migrate-dev` was last run. It is used
+-- to help warn the user of instances they may need to migrate.
```
```suggestion
-- to help warn the user of instances when they may need to migrate.
```

Martinsos commented on waspc/src/Wasp/Generator/DbGenerator/Common.hs:
```diff
@@ -0,0 +1,62 @@
+
+-- | This file represents the checksum of schema.prisma
+-- at the point at which `prisma generate` was last run. It is used
+-- to know if we need to regenerate during generation or not.
```
```suggestion
-- to know if we need to regenerate schema.prisma during web app generation or not.
```

Martinsos commented on waspc/src/Wasp/Generator/DbGenerator/Jobs.hs:
```diff
@@ -1,18 +1,31 @@
+-- `--no-install` is the magic that causes this command to fail if npx cannot find it locally
+--    (either in node_modules, or globally in npx). We do not want to allow npx to ask
+--    a user without prisma to install the latest version.
+-- We also version pin to what we need so it won't accidentally find a different version globally
```
```suggestion
-- We also pin the version to what we need so it won't accidentally find a different version globally
```

Martinsos commented on waspc/src/Wasp/Generator/DbGenerator/Jobs.hs:
```diff
@@ -27,28 +40,105 @@ migrateDev projectDir = do
+
+-- | Runs the original job, and if there is an error, it attempts recovery by running the second job
+-- before retrying the original one last time.
+-- NOTE(shayne): We only want to forward one JobExit message to the external channel from this function, as callers will stop
+-- listening once they get it. This is why our job message forwarders do not send them and we do ourselves at the end.
```
This is implementation details, so I would move this comment inside the function body, at the appropriate location. Caller of the function doesn't care about this, but if you are looking at the implementation, of it, then you do.

Martinsos replied:
This part with JobExit is particularly tricky but also important to understand because it is an assumption that is not easy to be aware of.

Therefore I think it makes sense to be very verbose here about it, even if too much, to ensure reader understands the reasoning. You could explain that what we are doing here is really creating a job that might run multiple sub-jobs as part of it. It might run just one sub-job, or it might run 3 sub-jobs (first attempt, recovery, second attempt). Since job signals its end by sending JobExit message to its channel, that means we must make sure to not send JobExit if only one of the multiple sub-jobs we are running here is done, therefore we can't just forward all of sub-job messages of we would also forward JobExit too early. Therefore we suppress their JobExit messages and make sure to send a JobExit only when we are done with all the sub-jobs.
You could explain this in even more details, focusing particularly on situation with 3 sub-jobs, and explaining how we don't want the first 2 of them to return their job exit.

Martinsos commented on waspc/src/Wasp/Generator/DbGenerator/Jobs.hs:
```diff
@@ -27,28 +40,105 @@ migrateDev projectDir = do
+          }
+
+    -- Forwards all JobOutput messages to the output channel, but does not send JobExit
+    -- as multiple jobs may be forwarded to the same output channel. The caller is responsible
```
```suggestion
    -- as this job might be only part of a bigger job, therefore we might not want to signal end of the job yet. The caller is responsible
```

Martinsos commented on waspc/src/Wasp/Generator/DbGenerator/Jobs.hs:
```diff
@@ -27,28 +40,105 @@ migrateDev projectDir = do
+        JobExit _ -> return ()
+
+    -- Only forwards JobOutput messages for the retry attempt if it fails.
+    -- All first attempt output is ignored, and so is success on retry.
```
I would also mention here that we make sure to not emit JobExit.

Martinsos commented on waspc/src/Wasp/Generator/DbGenerator/Jobs.hs:
```diff
@@ -27,28 +40,105 @@ migrateDev projectDir = do
+-- before retrying the original one last time.
+-- NOTE(shayne): We only want to forward one JobExit message to the external channel from this function, as callers will stop
+-- listening once they get it. This is why our job message forwarders do not send them and we do ourselves at the end.
+retryJobOnErrorWith :: J.Job -> J.Job -> JobMessageForwardingStrategy -> J.Job
```
Idea: what about separating this function + its data structures into a separate module? It could be DbGenerator/Jobs/Retry.hs . It would make it clearer that this is a special, separate thing.

Final step would be to decouple it from Db and just make this a general Job utility. I understand the only problem here is obtaining that information about the Job type. I guess you could get it from one of the job messages -> possibly from the very last JobExit you get -> but you would need to capture those then. Easier solution is to get this value as a function parameter, but then you risk this value being different from the value in actual messages.
I understand this is probably too much work for now, and I agree we can leave this for possible future, but it might be beneficial to add a TODO right now saying that we could do this and what it would take to make it work.

Martinsos commented on waspc/src/Wasp/Generator/DbGenerator/Jobs.hs:
```diff
@@ -27,28 +40,105 @@ migrateDev projectDir = do
+    -- Only forwards JobOutput messages for the retry attempt if it fails.
+    -- All first attempt output is ignored, and so is success on retry.
+    silentUntilRetryFails :: Attempt -> Chan JobMessage -> Chan JobMessage -> IO ()
+    silentUntilRetryFails attempt fromChan toChan = silentUntilRetryFails' []
```
FYI, when defining inner functions like `silentUntilRetryFails'` that do nothing but are a bit more general and then you call them with some initial args, there is a convention in Haskell to name them `go`. That said, I don't think it is a very good convention, so this is probably better what you did.

Martinsos commented on waspc/src/Wasp/Generator/DbGenerator/Operations.hs:
```diff
@@ -60,3 +66,40 @@ copyMigrationsBackToSource genProjectRootDirAbs dbMigrationsDirInWaspProjectDirA
     waspMigrationsDir = SP.castDir dbMigrationsDirInWaspProjectDirAbs
     genProjectMigrationsDir = SP.castDir $ genProjectRootDirAbs SP.</> dbRootDirInProjectRootDir SP.</> dbMigrationsDirInDbRootDir
+
+writeDbSchemaChecksumToFile :: Path' Abs (Dir ProjectRootDir) -> Path' (Rel ProjectRootDir) File' -> IO ()
```
You might want to mention here that this assumes db schema has been generated, as it will read the db schema from the generated code.

Martinsos commented on waspc/src/Wasp/Generator/DbGenerator/Operations.hs:
```diff
@@ -60,3 +66,40 @@ copyMigrationsBackToSource genProjectRootDirAbs dbMigrationsDirInWaspProjectDirA
+
+generatePrismaClient :: Path' Abs (Dir ProjectRootDir) -> IO (Either String ())
+generatePrismaClient genProjectRootDirAbs = do
+  dbSchemaChecksumFileExists <- doesFileExist dbSchemaChecksumFp
+
+  if dbSchemaChecksumFileExists
+    then do
+      dbSchemaFileChecksum <- hexToString <$> checksumFromFilePath dbSchemaFp
+      dbChecksumFileContents <- readFile dbSchemaChecksumFp
+      if dbSchemaFileChecksum == dbChecksumFileContents then return $ Right () else generatePrismaClient'
+    else generatePrismaClient'
```
To make this piece of code a bit easier to read, I would consider rewriting it to
```hs
ifM wasCurrentSchemaAlreadyGenerated (return $ Right()) generatePrismaClient'

...


where
  wasCurrentSchemaAlreadyGenerated :: IO Bool
  wasCurrentSchemaAlreadyGenerated = ...
```

Martinsos replied:
btw I never before used `ifM` but I just thought "there must be some way to do if-then-else on monadic value and hoogle said there is ifM!

shayneczyzewski replied:
Given that this was an external library, and to not overly populate our Util.hs file, I decided to punt on this one for now. Additionally, when trying to make use of it to clean things up I felt it did not make it that much easier to follow (for me at least) with the fact that we have a file existence check before we can check the contents. That outer check either had to wrap the `ifM` (and move the read farther away from the FilePath check, which felt off) or build it into the helper, but then we didn't clean up the shape too much in the latter case. I appreciate the idea though and sorry if I misunderstood 👍🏻

Martinsos reviewed (approved):
@shayneczyzewski Great, this is it!
I left some more comments, but mostly just small refactorings. I already approved, so once you are happy with those feel free to merge.

shayneczyzewski commented on waspc/src/Wasp/Generator.hs:
```diff
@@ -77,3 +75,10 @@ writeDotWaspInfo dstDir = do
   let dstPath = dstDir </> [relfile|.waspinfo|]
   Data.Text.IO.writeFile (SP.toFilePath dstPath) (Data.Text.pack content)
+
+-- | This function operates on generated code, and thus assumes the file drafts were written to disk
```
Yeah, probably so; I will move it over 👍🏻
